iavl/iavl_range_proofs.go

```diff
@@ -0,0 +1,400 @@
+		return err
+	}
+	if proof.PathToKey == nil {
+		// If we don't have an existing key, we effectively have a proof of absence.
+		return verifyKeyAbsence(proof.Left, proof.Right)
+	}
+
+	if err := proof.KeyExistsProof.Verify(key, value, root); err != nil {
+		return err
+	}
+	if bytes.Equal(key, endKey) {
+		return nil
+	}
+	if proof.PathToKey.isRightmost() {
+		return nil
+	}
+	if proof.Right != nil &&
+		proof.PathToKey.isLeftAdjacentTo(proof.Right.Path) {
+		return nil
+	}
+
+	return ErrInvalidProof
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+// KeyRangeProof is proof that a range of keys does or does not exist.
+type KeyRangeProof struct {
+	RootHash   data.Bytes   `json:"root_hash"`
+	PathToKeys []*PathToKey `json:"paths"`
+
+	Left  *PathWithNode `json:"left"`
+	Right *PathWithNode `json:"right"`
+}
+
+// Verify that a range proof is valid.
+//
+// This method expects the same parameters passed to query the range.
+func (proof *KeyRangeProof) Verify(
+	startKey, endKey []byte, limit int, keys, values [][]byte, root []byte,
+) error {
+	if len(proof.PathToKeys) != len(keys) || len(values) != len(keys) {
+		return ErrInvalidInputs
+	}
+	if limit > 0 && len(keys) > limit {
+		return ErrInvalidInputs
+	}
+	if len(proof.PathToKeys) == 0 && proof.Left == nil && proof.Right == nil {
+		return ErrInvalidProof
+	}
+
+	// If startKey > endKey, reverse the keys and values, since our proofs are
+	// always in ascending order.
+	startKey, endKey, keys, values = reverseIfDescending(startKey, endKey, keys, values)
+
+	{
+		firstKey, lastKey := startKey, endKey
+		if limit > 0 && len(keys) == limit {
+			firstKey, lastKey = keys[0], keys[len(keys)-1]
+		}
+		if err := verifyPaths(proof.Left, proof.Right, firstKey, lastKey, root); err != nil {
+			return err
+		}
+	}
+
+	// If keys is empty, it means we have an empty range. This range
+	// can be between keys, or outside of the range of existing keys.
+	if len(keys) == 0 {
+		return verifyKeyAbsence(proof.Left, proof.Right)
+	}
+
+	if err := verifyNoMissingKeys(proof.paths()); err != nil {
+		return err
+	}
+
+	// If we've reached this point, it means our range isn't empty, and we have
+	// a list of keys.
+	for i, path := range proof.PathToKeys {
+		leafNode := IAVLProofLeafNode{KeyBytes: keys[i], ValueBytes: values[i]}
+		if err := path.verify(leafNode, root); err != nil {
+			return err
+		}
+	}
+
+	if !bytes.Equal(startKey, keys[0]) && proof.Left == nil {
+		if !proof.PathToKeys[0].isLeftmost() {
+			return errors.New("left path is nil and first inner path is not leftmost")
+		}
+	}
+
+	// If the right proof is nil and the limit wasn't reached, we have to verify that
+	// we're not missing any keys. Basically, if a key to the right is missing because
+	// we've reached the limit, then it's fine. But if the key count is smaller than
+	// the limit, we need a right proof to make sure no keys are missing.
+	if proof.Right == nil && len(keys) != limit {
```

Comment:
Looks very good.
However, what about limits and descending proofs?
In that case, wouldn't the left side be the part that was cut off by the limit?

Is there a test case for the reverse here?